[processor/filter] Introduce inferred-context

kaisecheng opened the pull request:
This PR continues the work from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39688 by @jaronoff97
- Carried over the original implementation
- Fixed error_mode test cases
- Fixed resource filtering
- Fixed empty condition checking
- Added context inference to Profiles
- Added support to flat and advance configuration styles

<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This PR adds the ability to do context inference for conditions for the filter processor. New config fields are introduced for the supported telemetry types and an isolated code path is introduced for ease of review and separation of concerns

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37904

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Unit, ran locally

<!--Describe the documentation added.-->
#### Documentation
TODO
<!--Please delete paragraphs that you did not use before submitting.-->

kaisecheng committed:
Introduce inferred-context for filterprocessor

This commit continues the work from PR #39688 by @jaronoff97
- Carried over the original implementation
- Fixed error_mode ignore test cases
- Fixed issues with resource filtering

Co-authored-by: jaronoff97 <jacobaronoff45@gmail.com>

kaisecheng committed:
Fixed empty condition checking

kaisecheng committed:
Remove redundant validation

kaisecheng committed:
Add context inference to Profiles

kaisecheng committed:
lint and test

edmocosta commented on processor/filterprocessor/internal/contextfilter/logs_test.go:
These tests (logs, metrics, traces, profiles) should be and follow the same `processor/filterprocessor/<signal>_test.go` strategy. Could you please move/adapt them into there? We're also missing a few tests scenarios that can be ported from the transform processor, for example, error mode (global and per configuration group), Inference per context/sub-context.

edmocosta commented on processor/filterprocessor/config.go:
```diff
@@ -46,6 +47,14 @@ type Config struct {
 
 	Profiles ProfileFilters `mapstructure:"profiles"`
 
+	MetricConditions []common.ContextConditions `mapstructure:"metric_conditions"`
```
We're probably gonna need something similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/transformprocessor/config.go#L80, so we can support both formats, basic (flat list of conditions) and advanced configurations (this struct).

edmocosta commented on processor/filterprocessor/config.go:
```diff
@@ -385,3 +401,80 @@ func (cfg *Config) Validate() error {
+	}
+
+	if len(cfg.MetricConditions) > 0 {
+		pc, err := common.NewMetricParserCollection(component.TelemetrySettings{Logger: zap.NewNop()}, common.WithMetricParser(filterottl.StandardMetricFuncs()), common.WithDataPointParser(filterottl.StandardDataPointFuncs()), common.WithMetricErrorMode(cfg.ErrorMode))
```
```suggestion
		pc, err := common.NewMetricParserCollection(component.TelemetrySettings{Logger: zap.NewNop()}, common.WithMetricParser(cfg.metricFunctions), common.WithDataPointParser(cfg.dataPointFunctions), common.WithMetricErrorMode(cfg.ErrorMode))
```

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng committed:
centralize Parser Collection creation

kaisecheng committed:
centralize Parser Collection creation

kaisecheng committed:
add support to flat configuration style

kaisecheng committed:
avoid logging ErrSkipProcessingData

kaisecheng committed:
refactor tests

kaisecheng committed:
Merge branch 'main' of upstream

kaisecheng committed:
Use pointer when passing TransformContext

kaisecheng committed:
add deprecated comment

kaisecheng committed:
rename package name

kaisecheng committed:
remove unused functions

kaisecheng committed:
Merge branch 'main' into filterprocessor_context_inference

kaisecheng committed:
disable lint for package name common

kaisecheng commented on processor/filterprocessor/internal/common/config.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+// ContextConditions is a wrapper struct for OTTL conditions.
+type ContextConditions struct {
```
Added `context` in conditions

edmocosta commented on processor/filterprocessor/config.go:
```diff
@@ -323,65 +422,176 @@ func (cfg *Config) Validate() error {
+	var errs error
 
 	if cfg.Traces.ResourceConditions != nil {
 		_, err := filterottl.NewBoolExprForResource(cfg.Metrics.ResourceConditions, cfg.resourceFunctions, ottl.PropagateError, component.TelemetrySettings{Logger: zap.NewNop()})
```
```suggestion
		_, err := filterottl.NewBoolExprForResource(cfg.Traces.ResourceConditions, cfg.resourceFunctions, ottl.PropagateError, component.TelemetrySettings{Logger: zap.NewNop()})
```

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng committed:
update error message

kaisecheng committed:
rename package name to condition

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

kaisecheng commented on processor/filterprocessor/internal/common/config.go:
```diff
@@ -0,0 +1,67 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//revive:disable:var-naming
```
renamed to `condition`. I recall the package name is suggested to use singular noun

kaisecheng committed:
make goporto

edmocosta reviewed (approved):
It looks good to me, thank you for working on this @kaisecheng! I wish we could use the parser collection for all OTTL settings, but unfortunately it's not currently feasible without degrading the existing performance. That said, I think the logic repetition is fine for now, and we can clean this up later when the old settings are removed. 
The next step is documenting the new format, similarly to the transform processor's docs.

@TylerHelmuth @evan-bradley @bogdandrutu, could you please have a look?

kaisecheng committed:
Update processor/filterprocessor/config.go

Co-authored-by: Edmo Vamerlatti Costa <11836452+edmocosta@users.noreply.github.com>

TylerHelmuth commented on processor/filterprocessor/config.go:
```diff
@@ -309,79 +318,281 @@ type ProfileFilters struct {
+//
+//	log_conditions:
+//	  - resource.attributes["key1"] == "value"
+//	  - resource.attributes["key2"] == "value"
```
Can you remind me how we handle choosing the context for the flat style? What happens if the user does:

```
resource.attributes["key1"] == "value"
datapoint.attributes["key1"] == "value"
span.attributes["key1"] == "value"
```

kaisecheng replied:
We pick the finest context that inferred in conditions for the configured signal.
Taking metrics as example, resource vs datapoint would result in datapoint as inferred context. However, when users config mixed signals, such as datapoint (metric) and span (trace), it should give error.

edmocosta replied:
That's a good question, and I'm wondering now if it makes sense to keep it as it's (to match the transform processor), or if the filter processor should treat each flat condition as a new config block, with its own context inference and target data?

Let me give an example:

```
  trace_conditions:
    - span.span_id == SpanID("eee19b7ec3c1b173")
    - resource.attributes["key1"] == "value"
    - span.span_id == SpanID("eee19b7ec3c1b173") and spanevent.attributes["key1"] == "value"
```

Currently, the above config will infer the `spanevent` context, meaning that it will target the `span.events` entries, and **ALL** conditions must match to filter the data out. If any data matches only `resource.attributes["key1"] == "value"` or `span.span_id == SpanID("eee19b7ec3c1b173")` conditions, it won't be filtered out, which is a kind of `AND` relation between conditions, and is different from the non-inferred config ORed behavior. 

I understand it makes sense for the TP, where previous statements affects the execution of the next one, but I don't think it's the case for the FP. If we parse each condition independently, we would allow users to filter different parts of the data using a single ` trace_conditions` block, and the above config would:

`span.span_id == SpanID("eee19b7ec3c1b173")` -> infer `span` and remove all span that matches it
`resource.attributes["key1"] == "value"` ->  infer `resource` and remove all resource spans that matches it
`span.span_id == SpanID("eee19b7ec3c1b173") and spanevent.attributes["key1"] == "value"` -> infer `spanevent` and remove all span events that matches it.

WDYT? @TylerHelmuth @evan-bradley?

kaisecheng replied:
Agree this is a bug for filterprocessor. The problem is when the context is resolved to `spanevent` and the `spanevent` has no data, len(spenevent) == 0, the filter doesn't iterate the `spanevent`. Hence, ottl evaluation aren't executed and conditions for `resource` and `span` are not executed as well

TylerHelmuth reviewed (commented):
First quick pass complete. It looks very similar to transformprocessor implementation which is great. When I have more time later today I'll give it a more in depth review.

kaisecheng committed:
lint

TylerHelmuth commented on processor/filterprocessor/traces.go:
```diff
@@ -42,6 +45,23 @@ func newFilterSpansProcessor(set processor.Settings, cfg *Config) (*filterSpanPr
+		for _, cs := range cfg.TraceConditions {
+			metricConsumer, parseErr := pc.ParseContextConditions(cs)
+			errs = multierr.Append(errs, parseErr)
+			fsp.consumers = append(fsp.consumers, metricConsumer)
```
```suggestion
			traceConsumer, parseErr := pc.ParseContextConditions(cs)
			errs = multierr.Append(errs, parseErr)
			fsp.consumers = append(fsp.consumers, traceConsumer)
```

jaronoff97 reviewed (approved):
Looks great thanks for taking this over the finish line :bow:

kaisecheng committed:
Update processor/filterprocessor/traces.go

Co-authored-by: Tyler Helmuth <12352919+TylerHelmuth@users.noreply.github.com>

kaisecheng commented:
Many thanks to @edmocosta for the thorough reviews and helpful suggestions. Also thanks to @TylerHelmuth for the fast turnaround, and to @jaronoff97 for taking the initiative. Really appreciate everyone’s time and effort on this PR! 🙏

CI is green. This is ready to merge.

edmocosta reviewed (changes requested):
Let's discuss https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44543#discussion_r2693707798 before merging.

kaisecheng committed:
Handle zero-record lower-context conditions

Fix an edge case where conditions involving both higher and lower contexts
were skipped entirely when the lower context had zero records, causing
higher-context conditions to be incorrectly ignored.

The fix parses each condition as an independent condition group, infers
context per group, and rewrites how the filter processor iterates over
signal records.

kaisecheng committed:
Update dependencies

kaisecheng committed:
Merge branch 'main' into filterprocessor_context_inference

kaisecheng committed:
Fix typo

kaisecheng committed:
Remove unnecessary iteration

kaisecheng commented:
@edmocosta  For the edge case in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44543/#discussion_r2693707798, I rewrote how the filterprocessor iterates over signal. In flat style, each condition is parsed as an independent condition group and its context is inferred independently. The data now iterates through all context levels and executes OTTL condition when the condition is defined for the corresponding context.

This is ready for review

kaisecheng committed:
Remove unnecessary iteration

edmocosta commented on processor/filterprocessor/internal/condition/logs.go:
```diff
@@ -0,0 +1,239 @@
+// Flat style: the context is empty and each condition is parsed independently.
+// Conditions are then aggregated by their inferred context (resource, scope, log).
+// The processor level error mode is used for all conditions.
+func (lpc *LogParserCollection) ParseContextConditions(contextConditions ContextConditions) (LogConditions, error) {
```
```suggestion
func (lpc *LogParserCollection) ParseContextConditions(contextConditions ContextConditions) (LogsConsumer, error) {
```

It seems we only need the `LogsConsumer` outside this package, so we could replace it here with that and don't expose this internal struct.

edmocosta reviewed (commented):
Thank you @kaisecheng for working on this. I'm still reviewing it, but it looks good so far.  I've left a few initial suggestions so we can work in parallel. It also apply to the other signals :)

kaisecheng committed:
Remove CommonConditions

kaisecheng committed:
Rename & unexport

kaisecheng committed:
Refactor methods to constructor functions

kaisecheng committed:
Rename

kaisecheng committed:
Rename to common

kaisecheng committed:
Fix error mode for undefined context conditions

error mode in condition group should always take the precedence

kaisecheng committed:
Update comment

kaisecheng committed:
Merge branch 'main' into filterprocessor_context_inference

edmocosta committed:
Merge branch 'main' into filterprocessor_context_inference

edmocosta reviewed (approved):
LGTM! great work @kaisecheng!
@evan-bradley @TylerHelmuth could you please review it?

evan-bradley committed:
Merge branch 'main' into filterprocessor_context_inference

evan-bradley committed:
Merge branch 'main' into filterprocessor_context_inference

evan-bradley reviewed (commented):
Sorry @kaisecheng, I merged in main since I saw this PR was 51 commits behind, and it looks like things have shifted a bit since then and there are some build errors. Could you address those?

kaisecheng committed:
Update profile context to pointer

kaisecheng committed:
Update processor/filterprocessor/testdata/config_ottl.yaml

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

kaisecheng committed:
Merge branch 'main' into filterprocessor_context_inference

kaisecheng committed:
make schemagen

kaisecheng commented:
@evan-bradley CI is green now. Please have a look.

evan-bradley committed:
Merge branch 'main' into filterprocessor_context_inference

evan-bradley reviewed (approved):
Thanks @kaisecheng! Really appreciate your persistence through a long review cycle.

evan-bradley committed:
Fix bad merge

edmocosta committed:
Merge branch 'main' into filterprocessor_context_inference

kaisecheng committed:
make crosslink
